src/hub.ts

```
@@ -45,6 +45,7 @@ interface PluginRow {
   /**  Maps address to a socketIds */
   addressSubscriptions: Map<string, Set<number>>
   plugin: AddressPlugin
+  pluginReady: Promise<void>
```

We can't use a promise for this, because plugins can come & go multiple times. You'll need to tackle the state machine directly, instead of abstracting the logic inside a promise, which can only trigger once.

Just go ahead and subscribe to all the callback as soon as the hub boots up - there is no need to wait for the plugin to be connected. You can safely subscribe to the `update` event immediately - it just won't fire until the plugin connects.

Then we should probably add a boolean property to the `AddressPlugin` type, `.connected`. Then, if the core sends a `subscribe` message, but the `.connected` flag isn't set on the matching plugin, we can just return a subscription failure. If the core sends a `subscribe` event, and `connected` is true, then we can return success. There is no race condition, because if the plugin disconnects, the disconnection message will always come after the subscription-success message.

---

This has been since refactored out because we need multiple blockbook connections... Just pushed the update today.